File: android/src/main/java/com/livekit/lynx/audio/LynxAudioModule.kt
// ─────────────────────────────────────────────────────────────────────────────
// @livekit/lynx-webrtc — android/.../audio/LynxAudioModule.kt
// Port of LivekitReactNativeModule audio methods to Lynx.
// Replaces ReactContext → LynxContext, RCTDeviceEventEmitter → sendGlobalEvent.
// ─────────────────────────────────────────────────────────────────────────────

package com.livekit.lynx.audio

import android.content.Context
import com.livekit.lynx.internal.TrackRegistry
import com.lynx.jsbridge.LynxMethod
import com.lynx.jsbridge.LynxModule
import com.lynx.react.bridge.Callback
import com.lynx.tasm.behavior.LynxContext
import org.json.JSONObject
import org.webrtc.AudioTrack
import org.webrtc.AudioTrackSink
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import kotlin.time.Duration.Companion.milliseconds

private const val LOCAL_PC_ID = -1

class LynxAudioModule(context: Context) : LynxModule(context) {

    private val lynxCtx get() = mContext as LynxContext
    private val sinkManager = LynxAudioSinkManager()

    @LynxMethod
    fun createVolumeProcessor(pcId: Double, trackId: String, callback: Callback) {
        val processor = LynxVolumeProcessor { volume, tag ->
            lynxCtx.sendGlobalEvent(
                "LK_VOLUME_PROCESSED",
                JSONObject().apply { put("volume", volume); put("id", tag) }.toString()
            )
        }
        val tag = sinkManager.register(processor)
        processor.reactTag = tag
        attachSink(processor, pcId.toInt(), trackId)
        callback.invoke(null, tag)
    }

    @LynxMethod
    fun deleteVolumeProcessor(reactTag: String, pcId: Double, trackId: String, callback: Callback) {
        detachSink(reactTag, pcId.toInt(), trackId)
        sinkManager.unregister(reactTag)
        callback.invoke(null, null)
    }

    @LynxMethod
    fun createMultibandVolumeProcessor(optionsJson: String, pcId: Double, trackId: String, callback: Callback) {
        val opts = JSONObject(optionsJson)
        val bands    = opts.optInt("bands", 5)
        val minFreq  = opts.optDouble("minFrequency", 1000.0).toFloat()
        val maxFreq  = opts.optDouble("maxFrequency", 8000.0).toFloat()
        val interval = opts.optDouble("updateInterval", 40.0)

        val processor = LynxMultibandVolumeProcessor(
            minFrequency = minFreq,
            maxFrequency = maxFreq,
            barCount = bands,
            interval = interval.milliseconds
        ) { magnitudes, tag ->
            lynxCtx.sendGlobalEvent(
                "LK_MULTIBAND_PROCESSED",
                JSONObject().apply {
                    put("magnitudes", magnitudes.toList())
                    put("id", tag)
                }.toString()
            )
        }
        val tag = sinkManager.register(processor)
        processor.reactTag = tag
        attachSink(processor, pcId.toInt(), trackId)
        processor.start()
        callback.invoke(null, tag)
    }

    @LynxMethod
    fun deleteMultibandVolumeProcessor(reactTag: String, pcId: Double, trackId: String, callback: Callback) {
        val sink = sinkManager.getSink(reactTag)
        detachSinkObject(sink, pcId.toInt(), trackId)
        sinkManager.unregister(reactTag)
        (sink as? LynxMultibandVolumeProcessor)?.release()
        callback.invoke(null, null)
    }

    @LynxMethod
    fun createAudioSinkListener(pcId: Double, trackId: String, callback: Callback) {
        val processor = LynxAudioSinkProcessor { data, tag ->
            lynxCtx.sendGlobalEvent(
                "LK_AUDIO_DATA",
                JSONObject().apply { put("data", data); put("id", tag) }.toString()
            )
        }
        val tag = sinkManager.register(processor)
        processor.reactTag = tag
        attachSink(processor, pcId.toInt(), trackId)
        callback.invoke(null, tag)
    }

    @LynxMethod
    fun deleteAudioSinkListener(reactTag: String, pcId: Double, trackId: String, callback: Callback) {
        detachSink(reactTag, pcId.toInt(), trackId)
        sinkManager.unregister(reactTag)
        callback.invoke(null, null)
    }

    @LynxMethod
    fun setDefaultAudioTrackVolume(volume: Double, callback: Callback) {
        LynxWebRTCDefaults.defaultRemoteAudioVolume = volume
        callback.invoke(null, null)
    }

    // ── Attach / detach helpers ───────────────────────────────────────────────

    private fun attachSink(sink: AudioTrackSink, pcId: Int, trackId: String) {
        if (pcId == LOCAL_PC_ID) {
            LynxAudioRecordDispatcher.registerSink(sink)
        } else {
            (TrackRegistry.getAudioTrack(pcId, trackId) as? AudioTrack)
                ?.addSink(sink)
                ?: android.util.Log.w("LynxAudioModule", "Audio track not found: pcId=$pcId trackId=$trackId")
        }
    }

    private fun detachSink(reactTag: String, pcId: Int, trackId: String) {
        val sink = sinkManager.getSink(reactTag) ?: return
        detachSinkObject(sink, pcId, trackId)
    }

    private fun detachSinkObject(sink: AudioTrackSink?, pcId: Int, trackId: String) {
        if (sink == null) return
        if (pcId == LOCAL_PC_ID) {
            LynxAudioRecordDispatcher.unregisterSink(sink)
        } else {
            (TrackRegistry.getAudioTrack(pcId, trackId) as? AudioTrack)?.removeSink(sink)
        }
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// LynxAudioSinkManager — thread-safe registry
// ─────────────────────────────────────────────────────────────────────────────

class LynxAudioSinkManager {
    private val sinks = ConcurrentHashMap<String, AudioTrackSink>()

    fun register(sink: AudioTrackSink): String {
        val tag = UUID.randomUUID().toString()
        sinks[tag] = sink
        return tag
    }

    fun unregister(tag: String) { sinks.remove(tag) }
    fun getSink(tag: String): AudioTrackSink? = sinks[tag]
}

// ─────────────────────────────────────────────────────────────────────────────
// LynxWebRTCDefaults
// ─────────────────────────────────────────────────────────────────────────────

object LynxWebRTCDefaults {
    @Volatile var defaultRemoteAudioVolume: Double = 1.0
}

// ─────────────────────────────────────────────────────────────────────────────
// LynxAudioRecordDispatcher — dispatches local mic samples to registered sinks
// Port of AudioRecordSamplesDispatcher from the RN SDK
// ─────────────────────────────────────────────────────────────────────────────

object LynxAudioRecordDispatcher : org.webrtc.audio.JavaAudioDeviceModule.SamplesReadyCallback {

    private val sinks = java.util.Collections.synchronizedSet(mutableSetOf<AudioTrackSink>())

    fun registerSink(sink: AudioTrackSink)   { sinks.add(sink) }
    fun unregisterSink(sink: AudioTrackSink) { sinks.remove(sink) }

    override fun onWebRtcAudioRecordSamplesReady(
        samples: org.webrtc.audio.JavaAudioDeviceModule.AudioSamples
    ) {
        val bitsPerSample = when (samples.audioFormat) {
            android.media.AudioFormat.ENCODING_PCM_8BIT  -> 8
            android.media.AudioFormat.ENCODING_PCM_16BIT -> 16
            android.media.AudioFormat.ENCODING_PCM_FLOAT -> 32
            else -> 16
        }
        val numFrames = samples.sampleRate / 100
        val ts = android.os.SystemClock.elapsedRealtime()
        val snapshot = synchronized(sinks) { sinks.toSet() }
        for (sink in snapshot) {
            val buf = java.nio.ByteBuffer.wrap(samples.data)
            sink.onData(buf, bitsPerSample, samples.sampleRate, samples.channelCount, numFrames, ts)
        }
    }
}

File: ios/LynxAudioModule.swift
// ─────────────────────────────────────────────────────────────────────────────
// @livekit/lynx-webrtc — ios/LynxAudioModule.swift
// Lynx NativeModule — audio volume/multiband processors and audio sink.
// Port of LivekitReactNativeModule audio methods.
//
// Registration (LynxInitProcessor.m):
//   [globalConfig registerModule:LynxAudioModule.class];
// ─────────────────────────────────────────────────────────────────────────────

import Foundation
import WebRTC

// MARK: - LynxAudioModule

@objc(LynxAudioModule)
public final class LynxAudioModule: NSObject {

    @objc public var eventEmitter: LynxEventEmitter?

    private let rendererManager = LynxAudioRendererManager()

    // MARK: - Module registration

    @objc public static func name() -> String { "LynxAudioModule" }

    @objc public static func methodLookup() -> [String: String] {
        return [
            "createVolumeProcessor":          NSStringFromSelector(#selector(createVolumeProcessor(_:trackId:callback:))),
            "deleteVolumeProcessor":          NSStringFromSelector(#selector(deleteVolumeProcessor(_:pcId:trackId:callback:))),
            "createMultibandVolumeProcessor": NSStringFromSelector(#selector(createMultibandVolumeProcessor(_:pcId:trackId:callback:))),
            "deleteMultibandVolumeProcessor": NSStringFromSelector(#selector(deleteMultibandVolumeProcessor(_:pcId:trackId:callback:))),
            "createAudioSinkListener":        NSStringFromSelector(#selector(createAudioSinkListener(_:trackId:callback:))),
            "deleteAudioSinkListener":        NSStringFromSelector(#selector(deleteAudioSinkListener(_:pcId:trackId:callback:))),
            "setDefaultAudioTrackVolume":     NSStringFromSelector(#selector(setDefaultAudioTrackVolume(_:callback:))),
        ]
    }

    // MARK: - Volume Processor

    @objc func createVolumeProcessor(
        _ pcId: Double,
        trackId: String,
        callback: @escaping (String?, String?) -> Void
    ) {
        guard let emitter = eventEmitter else {
            callback("eventEmitter not set", nil)
            return
        }
        let renderer = LynxVolumeAudioRenderer(intervalMs: 40.0, eventEmitter: emitter)
        let tag = rendererManager.register(renderer)
        renderer.reactTag = tag
        attach(renderer: renderer, pcId: Int(pcId), trackId: trackId)
        callback(nil, tag)
    }

    @objc func deleteVolumeProcessor(
        _ tag: String,
        pcId: Double,
        trackId: String,
        callback: @escaping (String?, String?) -> Void
    ) {
        detach(rendererTag: tag, pcId: Int(pcId), trackId: trackId)
        rendererManager.unregister(forTag: tag)
        callback(nil, nil)
    }

    // MARK: - Multiband Volume Processor

    @objc func createMultibandVolumeProcessor(
        _ optionsJson: String,
        pcId: Double,
        trackId: String,
        callback: @escaping (String?, String?) -> Void
    ) {
        guard let emitter = eventEmitter else {
            callback("eventEmitter not set", nil)
            return
        }

        let opts = parseJSON(optionsJson) ?? [:]
        let bands = opts["bands"] as? Int ?? 5
        let minFreq = (opts["minFrequency"] as? NSNumber)?.floatValue ?? 1000
        let maxFreq = (opts["maxFrequency"] as? NSNumber)?.floatValue ?? 8000
        let interval = (opts["updateInterval"] as? NSNumber)?.floatValue ?? 40

        let renderer = LynxMultibandVolumeAudioRenderer(
            bands: bands,
            minFrequency: minFreq,
            maxFrequency: maxFreq,
            intervalMs: interval,
            eventEmitter: emitter
        )
        let tag = rendererManager.register(renderer)
        renderer.reactTag = tag
        attach(renderer: renderer, pcId: Int(pcId), trackId: trackId)
        callback(nil, tag)
    }

    @objc func deleteMultibandVolumeProcessor(
        _ tag: String,
        pcId: Double,
        trackId: String,
        callback: @escaping (String?, String?) -> Void
    ) {
        detach(rendererTag: tag, pcId: Int(pcId), trackId: trackId)
        rendererManager.unregister(forTag: tag)
        callback(nil, nil)
    }

    // MARK: - Audio Sink Listener

    @objc func createAudioSinkListener(
        _ pcId: Double,
        trackId: String,
        callback: @escaping (String?, String?) -> Void
    ) {
        guard let emitter = eventEmitter else {
            callback("eventEmitter not set", nil)
            return
        }
        let renderer = LynxAudioSinkRenderer(eventEmitter: emitter)
        let tag = rendererManager.register(renderer)
        renderer.reactTag = tag
        attach(renderer: renderer, pcId: Int(pcId), trackId: trackId)
        callback(nil, tag)
    }

    @objc func deleteAudioSinkListener(
        _ tag: String,
        pcId: Double,
        trackId: String,
        callback: @escaping (String?, String?) -> Void
    ) {
        detach(rendererTag: tag, pcId: Int(pcId), trackId: trackId)
        rendererManager.unregisterAndDetach(forTag: tag)
        callback(nil, nil)
    }

    // MARK: - Default volume

    @objc func setDefaultAudioTrackVolume(
        _ volume: Double,
        callback: @escaping (String?, String?) -> Void
    ) {
        // Stored globally — applied when new tracks are subscribed.
        LynxWebRTCDefaults.defaultRemoteAudioVolume = volume
        callback(nil, nil)
    }

    // MARK: - Private attach / detach

    private static let localPcId = -1

    private func attach(
        renderer: RTCAudioRenderer,
        pcId: Int,
        trackId: String
    ) {
        Task {
            if pcId == Self.localPcId {
                // Local mic — route through audio processing pipeline
                LynxAudioProcessingManager.shared.addLocalAudioRenderer(renderer)
            } else {
                guard let track = await TrackRegistry.shared.getAudioTrack(
                    pcId: pcId,
                    trackId: trackId
                ) else {
                    print("[LynxAudioModule] Audio track not found: pcId=\(pcId) trackId=\(trackId)")
                    return
                }
                track.add(renderer)
            }
        }
    }

    private func detach(
        rendererTag: String,
        pcId: Int,
        trackId: String
    ) {
        guard let renderer = rendererManager.renderer(forTag: rendererTag) else { return }
        Task {
            if pcId == Self.localPcId {
                LynxAudioProcessingManager.shared.removeLocalAudioRenderer(renderer)
            } else {
                guard let track = await TrackRegistry.shared.getAudioTrack(
                    pcId: pcId,
                    trackId: trackId
                ) else { return }
                track.remove(renderer)
            }
        }
    }

    // MARK: - JSON helper

    private func parseJSON(_ json: String) -> [String: Any]? {
        guard let data = json.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return nil }
        return obj
    }
}

// MARK: - LynxWebRTCDefaults

/// Simple global defaults shared between modules.
enum LynxWebRTCDefaults {
    static var defaultRemoteAudioVolume: Double = 1.0
}

// MARK: - LynxAudioRendererManager

/// Thread-safe registry of active RTCAudioRenderer instances.
final class LynxAudioRendererManager: @unchecked Sendable {

    private let lock = NSLock()
    private var renderers: [String: RTCAudioRenderer] = [:]

    func register(_ renderer: RTCAudioRenderer) -> String {
        let tag = UUID().uuidString
        lock.withLock { renderers[tag] = renderer }
        return tag
    }

    func unregister(forTag tag: String) {
        lock.withLock { _ = renderers.removeValue(forKey: tag) }
    }

    func unregisterAndDetach(forTag tag: String) {
        lock.withLock { _ = renderers.removeValue(forKey: tag) }
    }

    func renderer(forTag tag: String) -> RTCAudioRenderer? {
        lock.withLock { renderers[tag] }
    }
}
